### ISSUE TITLE
Reflectometry: Improvements

### ISSUE THREAD

**Issue by John-Holt-Tessella** [ISISComputingGroup/IBEX#4207]
The following is a list of feedback from CRISP tests. We should confirm Mandatory and Optional labeling (only I have done this). These would need to be completed for a user program on SURF (and therefore for demo on SURF at end of next cycle). I have arranged these into areas for easier discussion:

### JASCO Pump

1.	(M) JASCO OPI does not show alarms when not communicating
2.	(O) Jasco pump status is 0 or 1 not words like RUNNING/Stopped
3.	(M) Jasco error status string need to be on front panel
12.	(O) On Jasco: allow concentrations to be set without pumping. This allows user to check that values are being set without pumping stuff into the cell
54. (O) Jasco labview VI to start a timed run needs set all pressed after the timed button is pressed. This is strange. Also values seem to be set only after the set is pressed. (Using it was strange go and play with it and make it nicer). After we have finished the Genie script the Christy wrote needs to be updated so let him know.
55. (M) How is the Jasco pump going to be used on SECI. Currently it needs a mini inst started with the ioc running in it (I think). Is this the plan? If so we should place IBEX on all instruments that might need it so we don't need to set it up in the middle of the night.

### Changes for CRISP

1.	(M) Need a Laser Mode: detector is up and multi-detector is up
14.	(O) Need a parameter on multi detector in/out of the beam
    1. Not needed for SURF so optional
15.	(O) With the multi-detector in and point detector out need to be able to change the park position dependent on the location of the multi detector. If it is high then the point-detector needs to be parked low and vice versa
    1. This is an odd case on CRISP only

### General GUI changes

1.	(O) OPI pressing return after each number entry is annoying. It should at the very least go to the next field on the press of return
    1. This is hard so I am hoping it is optional but please push back if it is not
17.	(M) The advanced motor tab needs the name of the motor on its tab title
21.	(M) Apply changes is not being enabled on the DAE tab after run then abort.
25.	(O) First OPI load is slow
31.	(O) Motor table needs to reflect the precision set on the motor
32.	(O) Need to be able to determine more motor information in case of fault finding. They would like to see on the motor screen indication that motor is auto de-energised, encoders are on, offset, limits, error between SP & Val
    1. Suggested having a simple/advanced version of the motor table and this would be fine
35.	(O) Can we display the beam currents for TS1 and TS2 in large letter on the beam status graph

### IBEX Changes

1.	(M) If the motor is in `set` mode rather than `use` mode the setpoint (and readback) should be alarmed so that a user knows they should not be changing their positions.

### Reflectometer OPI and IOC changes
    
1. (M) Sample parameters should have a setpoint, readback new setpoint like all other parameters
7. (M) Need move button next to all setpoint, readback new setpoint triplets
    1. May do this as two versions of the screen one with not readback setpoint and no individual moves for the user, and one with everything.
8. (M) The detector height does not change when it is scanned over; this is because of the definition of theta, fix this though because we need to scan this variable.
9. (M) Reset axis speeds after move has finished (even on ctrl+c) [and 47] #4045
13.	(M) The S4Rotation appears not to be synchornised i.e. the rotation finished a long time before the rest of the motion.
16.	(M) On the reflectometry OPI indicate that an axis is moving with green background. Indicate that the setpoint and setpoint read back are not synchronised by turning the setpoint read back red. Indicate that the setpoint readback and readback are not in sync by colouring the readback red.
28.	(M) Why is super mirror put back into the beam on move after a restart of reflectometry server. Is this because we don't have initialisation yet?
36.	(M) Readback, setpoint readback and setpoint should be labelled
38. (M) We need to talk through how the setpoints are updated if lower level set points are set. We thought that it was ok just to set it from the high level and push down but it turns out this is weird for some parameters like the gaps if you set it on a low table and then move on the top table.
48. (M) The height offset is being reapplied on theta change in crisp. it should not be because it is not part of the mode. (One reason this might happen is that move on the beamline applies all setpoints to setpoint readbacks every time)
36. (M) Create a reflectometry view which has scripting window, graph. blocks, dashboard and reflectometry front panel (either from device screens or synoptic).

### Motors

1.	(M) Slit 3 Jaws gap did not move, the bottom jaw appears to be getting stuck at the bottom of its travel. Talk with Ian about parameters
18.	(M) The settings of the pizo jaws look weird have these been transferred correctly from SECI?
19.	(M) Can we choose any velocity for the pizo jaws (i.e. for synchronised moves) or can they only run at one speed talk with I.J. about this
    
### Scans Library Changes

1.	(M) Scan graphs should be plotted with units on both x and y, y axis should be labeled; ideally spec3/spec4 but I/I0 is fine
22.	(M) Set output directory for scan output file; this should be changeable with a command
41. (M) The fitting error should be displayed on the plots and returned in the fitting object because a user needs to know how accurate this number is (1dp vs 4dp)
42. (O) Blocks in a scan should not be case sensitive (ticket #4136 will add an upper case alias which can be used)
43. (M) The number of dp's on the ERF fit is too small in some cases (lets try and use the error to size this).
44. (O) Centre of mass fit line is blue and should be orange like other fits
49. (O) Think about whether we can and want to set a variable to the centre after a fit

### Plotting

1.	(M) Maximum of 6 plots is disconcerting; plot should be removed when limit is reached [Limit is actually 48]
34.	(M) Make clearing all plots easier to remember; perhaps a command alias
37.	(O)  Plot should refresh faster

### Scripting

1.	(M) Auto load needed libraries at the start of a python session
27.	(M) Create a scan over chopper procedure (see script on CRISP). But instead save each phase to a different period and save as file that can be looked at in Mantid (also be good to plot spectra)
39.	(M)  The help in contrast change needs to define the concentration as an array with 4 values in, with example.
50. (M) The title in the run scripts include HGaps but it uses the set version not values so that if you don't set them the title becomes `[None None None None]`
51. (M) The documentation of procedures contrast change does not contain units for flow
52. (M) Have used `set period` not `change period number` in various places. run angle and transmission should use a single period (and set this up before starting)
53. (M) Title of the experiment in runangle and transmission should contain all svgs and hvgs

### Complete
1.	(O) If Jaws are moving and reset, does it work? 
a.	Yes it seems fine. Note that if the lower jaw is moved beyond its limit but the upper jaw isn’t you can crash the jaws. This is bad, either there should be a limit on slit gap/height or setting slit wide set them to the maximum
23.	(O) We need to be able to fit data after scan, so we can try different fits
26.	(O) +1 for stability analysis in IBEX request (maybe added to blocks)
    1. Ticket 3149
29.	(M) Allow fitting to top-hat with a background  
    1. This is done from general.scans.fit import Gaussian, Erf, TopHat, CentreOfMass then fit=TopHat
30.	(M) Having multiple graphs is good, can we delete one graph. 
    1. This is done looks like matplotlib.pyplot.close(figure number) – document
45. (M) Double check that the backlash and retry are included in "done move".
    1. It does see https://epics.anl.gov/bcda/synApps/motor/R4-5/motorRecord.html DMOV
46. (M) On CRISP the readback and setpoint of the s4Height axis appear not to track in the archive. This seems like an archive problem check it. (Thursday 28/3/2019 12:09:40 -> 12:16:30)
    1. This is the same artifact caused by #3779 and will be fixed next release. I plotted the data for a point that was off and it looked like the fllowing, you can see the data from the end of the second should be displaced to the end of the previous second. Then it hits the set point.
![image](https://user-images.githubusercontent.com/7570055/55341906-70c4f500-549f-11e9-9032-7334f4be5e4a.png)

**Comment by ThomasLohnert** [comment]
Comment to keep track of items above vs. tickets created to address them. Each of these items should eventually have a ticket next to it, and tickets should be sensibly grouped/scoped.

### JASCO Pump

1.	https://github.com/ISISComputingGroup/IBEX/issues/4257 (M) JASCO OPI does not show alarms when not communicating
2.	https://github.com/ISISComputingGroup/IBEX/issues/4284 (O) Jasco pump status is 0 or 1 not words like RUNNING/Stopped
3.	https://github.com/ISISComputingGroup/IBEX/issues/4198 (M) Jasco error status string need to be on front panel
12.	https://github.com/ISISComputingGroup/IBEX/issues/4283 (O) On Jasco: allow concentrations to be set without pumping. This allows user to check that values are being set without pumping stuff into the cell
54. https://github.com/ISISComputingGroup/IBEX/issues/4280 (O) Jasco labview VI to start a timed run needs set all pressed after the timed button is pressed. This is strange. Also values seem to be set only after the set is pressed. (Using it was strange go and play with it and make it nicer). After we have finished the Genie script the Christy wrote needs to be updated so let him know. 
55. https://github.com/ISISComputingGroup/IBEX/issues/4243 https://github.com/ISISComputingGroup/IBEX/issues/4242 (M) How is the Jasco pump going to be used on SECI. Currently it needs a mini inst started with the ioc running in it (I think). Is this the plan? If so we should place IBEX on all instruments that might need it so we don't need to set it up in the middle of the night.

### Changes for CRISP

1.	(M) Need a Laser Mode: detector is up and multi-detector is up
14.	(O) Need a parameter on multi detector in/out of the beam
    1. Not needed for SURF so optional
15.	(O) With the multi-detector in and point detector out need to be able to change the park position dependent on the location of the multi detector. If it is high then the point-detector needs to be parked low and vice versa
    1. This is an odd case on CRISP only

### General GUI changes

1.	(O) OPI pressing return after each number entry is annoying. It should at the very least go to the next field on the press of return
    1. This is hard so I am hoping it is optional but please push back if it is not
17.	(M) The advanced motor tab needs the name of the motor on its tab title
21.	(M) Apply changes is not being enabled on the DAE tab after run then abort.
25.	(O) First OPI load is slow
31.	(O) Motor table needs to reflect the precision set on the motor
32.	(O) Need to be able to determine more motor information in case of fault finding. They would like to see on the motor screen indication that motor is auto de-energised, encoders are on, offset, limits, error between SP & Val
    1. Suggested having a simple/advanced version of the motor table and this would be fine
35.	(O) Can we display the beam currents for TS1 and TS2 in large letter on the beam status graph

### IBEX Changes

1.	(M) If the motor is in `set` mode rather than `use` mode the setpoint (and readback) should be alarmed so that a user knows they should not be changing their positions.

### Reflectometer OPI and IOC changes
    
1. (M) Sample parameters should have a setpoint, readback new setpoint like all other parameters
7. (M) Need move button next to all setpoint, readback new setpoint triplets
    1. May do this as two versions of the screen one with not readback setpoint and no individual moves for the user, and one with everything.
8. (M) The detector height does not change when it is scanned over; this is because of the definition of theta, fix this though because we need to scan this variable.
9. (M) Reset axis speeds after move has finished (even on ctrl+c) [and 47] #4045
13.	(M) The S4Rotation appears not to be synchornised i.e. the rotation finished a long time before the rest of the motion.
16.	(M) On the reflectometry OPI indicate that an axis is moving with green background. Indicate that the setpoint and setpoint read back are not synchronised by turning the setpoint read back red. Indicate that the setpoint readback and readback are not in sync by colouring the readback red.
28.	(M) Why is super mirror put back into the beam on move after a restart of reflectometry server. Is this because we don't have initialisation yet?
36.	(M) Readback, setpoint readback and setpoint should be labelled
38. (M) We need to talk through how the setpoints are updated if lower level set points are set. We thought that it was ok just to set it from the high level and push down but it turns out this is weird for some parameters like the gaps if you set it on a low table and then move on the top table.
48. (M) The height offset is being reapplied on theta change in crisp. it should not be because it is not part of the mode. (One reason this might happen is that move on the beamline applies all setpoints to setpoint readbacks every time)
36. (M) Create a reflectometry view which has scripting window, graph. blocks, dashboard and reflectometry front panel (either from device screens or synoptic).

### Motors

1.	(M) Slit 3 Jaws gap did not move, the bottom jaw appears to be getting stuck at the bottom of its travel. Talk with Ian about parameters
18.	(M) The settings of the pizo jaws look weird have these been transferred correctly from SECI?
19.	(M) Can we choose any velocity for the pizo jaws (i.e. for synchronised moves) or can they only run at one speed talk with I.J. about this
    
### Scans Library Changes

1.	(M) Scan graphs should be plotted with units on both x and y, y axis should be labeled; ideally spec3/spec4 but I/I0 is fine
22.	(M) Set output directory for scan output file; this should be changeable with a command
41. (M) The fitting error should be displayed on the plots and returned in the fitting object because a user needs to know how accurate this number is (1dp vs 4dp)
42. (O) Blocks in a scan should not be case sensitive (ticket #4136 will add an upper case alias which can be used)
43. (M) The number of dp's on the ERF fit is too small in some cases (lets try and use the error to size this).
44. (O) Centre of mass fit line is blue and should be orange like other fits
49. 	(O) Think about whether we can and want to set a variable to the centre after a fit

### Plotting

1.	(M) Maximum of 6 plots is disconcerting; plot should be removed when limit is reached [Limit is actually 48]
34.	(M) Make clearing all plots easier to remember; perhaps a command alias
37.	(O)  Plot should refresh faster


### Scripting

1.	(M) Auto load needed libraries at the start of a python session
27.	(M) Create a scan over chopper procedure (see script on CRISP). But instead save each phase to a different period and save as file that can be looked at in Mantid (also be good to plot spectra)
39.	(M)  The help in contrast change needs to define the concentration as an array with 4 values in, with example.
50. (M) The title in the run scripts include HGaps but it uses the set version not values so that if you don't set them the title becomes `[None None None None]`
51. (M) The documentation of procedures contrast change does not contain units for flow
52. (M) Have used `set period` not `change period number` in various places. run angle and transmission should use a single period (and set this up before starting)
53. (M) Title of the experiment in runangle and transmission should contain all svgs and hvgs

**Comment by Tom-Willemsen** [comment]
Ideally @John-Holt-Tessella should review this - maybe with someone else so that more people know about reflectometry.
